### PR TITLE
Mention implementers of unsatisfied trait

### LINKED ISSUE #87437: Using str::contains inside of a filter leads to confusing error message.
I've been trying to learn some Rust by doing CodeWars. I've spent some time being stuck on this diagnostic. One of the easiest exercises there is to count vowels in a string. First I did it with a loop, but then got stuck trying to rewrite is as a filter.

Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=785ded94c58b6d9a3a8acd16145a4970

```rust
fn get_vowel_count(string: &str) -> usize {
        string
        .chars()
        .filter(|c| "aeiou".contains(c))
        .count()
}
```

The current output is:

```
error[E0277]: expected a `Fn<(char,)>` closure, found `char`
 --> src/lib.rs:4:38
  |
4 |         .filter(|c| "aeiou".contains(c))
  |                                      ^ expected an `Fn<(char,)>` closure, found `char`
  |
  = help: the trait `Fn<(char,)>` is not implemented for `char`
  = note: required because of the requirements on the impl of `FnOnce<(char,)>` for `&char`
  = note: required because of the requirements on the impl of `Pattern<'_>` for `&char`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `challenge`
```

Ideally the output should not lead with "expected an Fn" and also help me to add a missing ampersand.

### LINKED ISSUE #90970: Misleading/Incomplete errors involving Pattern
Given the following code: 

```rust
pub fn strip_lf(s: &str) -> &str {
    s.strip_suffix(b'\n').unwrap_or(s)
}
```

The current output is:

```
error[E0277]: expected a `FnMut<(char,)>` closure, found `u8`
 --> src/lib.rs:2:20
  |
2 |     s.strip_suffix(b'\n').unwrap_or(s)
  |                    ^^^^^ expected an `FnMut<(char,)>` closure, found `u8`
  |
  = help: the trait `FnMut<(char,)>` is not implemented for `u8`
  = note: required because of the requirements on the impl of `Pattern<'_>` for `u8`
```

`Pattern<'a>` has implementations for:
- `char`
- `&str`
- `&String`
- `&[char]`
- `&&str`
- `F` where `F: FnMut(char) -> bool`

While it is true that u8 doesn't match any of the non-generic impls, and that it doesn't have an impl for `FnMut<(char,)> -> bool`, it's sad that the error message doesn't mention that those other possibilities exist.
(Also, the error message doesn't mention the closure return type)

### PR DESCRIPTION
**Pull request by estebank**
When encountering an unsatisfied trait bound, if there are no other
suggestions, mention all the types that *do* implement that trait:

```
error[E0277]: the trait bound `f32: Foo` is not satisfied
  --> $DIR/impl_wf.rs:22:6
   |
LL | impl Baz<f32> for f32 { }
   |      ^^^^^^^^ the trait `Foo` is not implemented for `f32`
   |
   = help: the trait `Foo` is implemented for `i32`
note: required by a bound in `Baz`
  --> $DIR/impl_wf.rs:18:31
   |
LL | trait Baz<U: ?Sized> where U: Foo { }
   |                               ^^^ required by this bound in `Baz`
```
```
error[E0277]: the trait bound `u32: Foo` is not satisfied
  --> $DIR/associated-types-path-2.rs:29:5
   |
LL |     f1(2u32, 4u32);
   |     ^^ the trait `Foo` is not implemented for `u32`
   |
   = help: the trait `Foo` is implemented for `i32`
note: required by a bound in `f1`
  --> $DIR/associated-types-path-2.rs:13:14
   |
LL | pub fn f1<T: Foo>(a: T, x: T::A) {}
   |              ^^^ required by this bound in `f1`
```

Suggest dereferencing in more cases.

Fix #87437, fix #90970.

### CODE REVIEW

**Comment by fmease** [comment]
Does this fix #90970?

**Comment by estebank** [comment]
As is, it does not:

```
error[E0277]: expected a `FnMut<(char,)>` closure, found `u8`
 --> f23.rs:2:20
  |
2 |     s.strip_suffix(b'\n').unwrap_or(s)
  |       ------------ ^^^^^ expected an `FnMut<(char,)>` closure, found `u8`
  |       |
  |       required by a bound introduced by this call
  |
  = help: the trait `FnMut<(char,)>` is not implemented for `u8`
```

Looking if there's a straightforward change that can be done.

**Comment by estebank** [comment]
Update: [now it does cover it, slightly](https://github.com/rust-lang/rust/pull/91873/files#diff-4f26c89466dac0f7a959d74b672577fec3a66f6468cd363b618eea9eb71267fd).

**Comment by fmease** [comment]
> What do you think?

I think, it's already an improvement that it lists those other implementations and thus likely clears up much faster the potential confusion of the user which arises from the incorrect mention of `FnMut` (instead of `Pattern`). So thank you very much! :heart:

I first faced this error message a few years ago when I wasn't that familiar with Rust yet and I remember it took me a quite a bit of time to figure out what was going on. So any steps towards fixing this is great!

> I would like to iterate a bit further (and report the error higher in the obligation chain so we only talk about Pattern and not FnMut), but that will be better served by a separate PR.

Sure! I hope you (or someone else of course) can figure out how to suppress, really, references to `Fn*` obligations in certain situations. From past experiences, I know that sometimes for some reason `Fn*` impls  like to “take over” certain error messages. I don't have a concrete example at hand right now, though (apart from `Pattern` obviously).

**Comment by michaelwoerister** [comment]
Sorry for the delay. I think it might be better if someone from wg-diagnostics reviewed this.

r? rust-lang/diagnostics

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 1]
```diff
@@ -305,12 +307,12 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
 
+                        let def_id = trait_ref.def_id();
                         let OnUnimplementedNote { message, label, note, enclosing_scope } =
                             self.on_unimplemented_note(trait_ref, &obligation);
                         let have_alt_message = message.is_some() || label.is_some();
-                        let is_try_conversion = self.is_try_conversion(span, trait_ref.def_id());
-                        let is_unsize =
-                            { Some(trait_ref.def_id()) == self.tcx.lang_items().unsize_trait() };
+                        let is_try_conversion = self.is_try_conversion(span, def_id);
+                        let is_unsize = { Some(def_id) == self.tcx.lang_items().unsize_trait() };
```
Comment (davidtwco) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:315]: Are these parentheses redundant?

File: compiler/rustc_errors/src/diagnostic.rs [hunk 2]
```diff
@@ -293,6 +293,11 @@ impl Diagnostic {
         self
     }
 
+    pub fn highlighted_help(&mut self, msg: Vec<(String, Style)>) -> &mut Self {
```
Comment (davidtwco) [inline compiler/rustc_errors/src/diagnostic.rs:296]: Perhaps add a documentation comment describing why someone might want to use `highlighted_help` when implementing a diagnostic?

File: compiler/rustc_middle/src/ty/mod.rs [hunk 3]
```diff
@@ -2049,6 +2049,11 @@ impl<'tcx> TyCtxt<'tcx> {
         })
     }
 
+    /// If the given DefId belongs to a trait that was automatically derived, returns `true`.
```
Comment (davidtwco) [inline compiler/rustc_middle/src/ty/mod.rs:2052]:
```suggestion
    /// If the given `DefId` belongs to a trait that was automatically derived, returns `true`.
```
nit: backticks

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 4]
```diff
@@ -1478,23 +1489,72 @@ impl<'a, 'tcx> InferCtxtPrivExt<'tcx> for InferCtxt<'a, 'tcx> {
...
+                .filter(|&def_id| {
+                    self.tcx.impl_polarity(def_id) != ty::ImplPolarity::Negative
+                        || self.tcx.is_builtin_derive(def_id)
+                })
+                .filter_map(|def_id| self.tcx.impl_trait_ref(def_id))
+                // Avoid mentioning type parameters.
+                .filter(|trait_ref| !matches!(trait_ref.self_ty().kind(), ty::Param(_)))
+                .map(|trait_ref| format!("\n  {}", trait_ref.self_ty()))
+                .collect();
+            normalized_impl_candidates.sort();
```
Comment (davidtwco) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1526]: The logic for emitting the actual diagnostic after `sort`/`dedup` feels quite similar in this case and the diagnostic emitted below, do you think we could deduplicate a little bit here?

Maybe this?

```rust
let impl_candidates = if impl_candidates.is_empty() { 
  /* what you do now up to this line */
 } else { 
  impl_candidates 
};
```

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 5]
```diff
@@ -2225,16 +2233,77 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
...
+
+                        // Mention *all* the applicable `impl`s, the user might have meant to use one of them.
+                        let mut items: Vec<_> = def_ids
+                            .iter()
+                            .filter_map(|&def_id| self.tcx.impl_trait_ref(def_id))
+                            .map(|trait_ref| format!("\n  {}", trait_ref.self_ty()))
+                            .collect();
+                        items.sort();
+                        items.dedup();
+                        if !items.is_empty() {
```
Comment (davidtwco) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:2293]: I wonder if we could use a helper function for taking some list, sorting, deduplicating and writing a diagnostic with the first element or the first N elements?

**Review by davidtwco — CHANGES_REQUESTED** [review]
I like the diagnostic change, couple comments on the implementation where we might be able to deduplicate things a bit.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs (hunk 4 above)
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs]: done

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 6]
```diff
@@ -1571,24 +1649,95 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
...
+                ]);
+                return true;
+            }
+            let trait_ref = TraitRef::identity(self.tcx, candidates[0].def_id);
+            // Check if the trait is the same in all cases. If so, we'll only show the type.
+            // FIXME: there *has* to be a better way!
+            let mut traits: Vec<_> = candidates
+                .iter()
+                .map(|c| format!("{}", c).split(" as ").last().unwrap().to_string())
+                .collect();
```
Comment (estebank) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1681]: Calling this out: it's a hack and I was too sleepy to come up with a replacement.

I think `print_only_trait_path` (as opposed to `print_only_trait_name`) will work?

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs (hunk 1 above)
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs]: Yes, removed.

File: compiler/rustc_errors/src/diagnostic.rs (hunk 2 above)
Comment (estebank) [reply compiler/rustc_errors/src/diagnostic.rs:363]: Added docstring

File: compiler/rustc_middle/src/ty/mod.rs (hunk 3 above)
Comment (estebank) [reply compiler/rustc_middle/src/ty/mod.rs]: Changed

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs (hunk 5 above)
Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs]: Moved both cases in this file to a single closure, but yes, I'm sure there are three or four other impls of this algorithm in the codebase. I'll leave that for a later follow up.

**Review by davidtwco — APPROVED** [review]
r=me unless you want to squash the commits first

**Pull request merged**